### PR TITLE
Add header support for middleware

### PR DESCRIPTION
**Pull request by namusyaka**
ref #1069 
This adds the `header` method for activating it in our middleware.
Could you review the pull request?

### CODE REVIEW

File: lib/grape/dsl/inside_route.rb [hunk 1]
```diff
@@ -138,16 +138,20 @@ def status(status = nil)
...
       end
 
-      # Set an individual header or retrieve
-      # all headers that have been set.
-      def header(key = nil, val = nil)
-        if key
-          val ? @header[key.to_s] = val : @header.delete(key.to_s)
-        else
-          @header
+      module HeaderMethods
```
Comment (dblock) [inline lib/grape/dsl/inside_route.rb]: Should be in a separate file probably. And I don't love the name "HeaderMethods", maybe just "Headers"?

File: lib/grape/middleware/base.rb [hunk 2]
```diff
@@ -27,7 +29,21 @@ def call!(env)
...
         end
-        after_response || @app_response
+
+        response = after_response || @app_response
+        reflect_headers response
+        response
+      end
+
+      def reflect_headers(response)
+        return unless header.is_a?(Hash)
```
Comment (dblock) [inline lib/grape/middleware/base.rb]: What else can `header` be? I think a check for `nil` is more appropriate, so `unless header`. Also alias `header` to `headers`, when we check on a collection of headers we should be doing `headers`.

File: lib/grape/middleware/base.rb [hunk 3]
```diff
@@ -27,7 +29,21 @@ def call!(env)
...
+        response_header =
+          case response
+          when Rack::Response then response.headers
+          when Array          then response[1]
+          end
+        response_header.merge!(header)
+        header.clear
```
Comment (dblock) [inline lib/grape/middleware/base.rb]: Why is this necessary?

File: lib/grape/dsl/inside_route.rb (hunk 1 above)
Comment (namusyaka) [inline lib/grape/dsl/inside_route.rb]: Agreed. I'll fix it.

File: lib/grape/middleware/base.rb (hunk 2 above)
Comment (namusyaka) [inline lib/grape/middleware/base.rb]: Nope, there is a possibility of inserting `Rack::Accept::MediaType` into `@headers`.
See https://github.com/ruby-grape/grape/blob/master/lib/grape/middleware/versioner/header.rb#L69

However, aliasing seems reasonable to me. So I'd like to fix it.

File: lib/grape/middleware/base.rb (hunk 3 above)
Comment (namusyaka) [inline lib/grape/middleware/base.rb]: Right. My Bad. Thanks!

**Comment by namusyaka** [comment]
Fixed and force pushed.

File: lib/grape/middleware/base.rb [hunk 4]
```diff
@@ -27,7 +31,20 @@ def call!(env)
...
           after_response = after
         end
-        after_response || @app_response
+
+        response = after_response || @app_response
+        reflect_headers response
+        response
+      end
+
+      def reflect_headers(response)
```
Comment (dblock) [inline lib/grape/middleware/base.rb]: Should this be private?

Maybe rename it to `merge_headers`?

File: lib/grape/middleware/base.rb [hunk 5]
```diff
@@ -27,7 +31,20 @@ def call!(env)
...
+        return unless headers.is_a?(Hash)
+        response_headers =
+          case response
+          when Rack::Response then response.headers
+          when Array          then response[1]
+          end
+        response_headers.merge!(headers)
```
Comment (dblock) [inline lib/grape/middleware/base.rb]: If the API developer made a mistake and `response` is neither, we get a `nil`, I would stay defensive and do `merge!() ... if response_headers`.

File: lib/grape/middleware/base.rb [hunk 6]
```diff
@@ -27,7 +31,20 @@ def call!(env)
...
-        after_response || @app_response
+
+        response = after_response || @app_response
+        reflect_headers response
+        response
+      end
+
+      def reflect_headers(response)
+        return unless headers.is_a?(Hash)
+        response_headers =
```
Comment (dblock) [inline lib/grape/middleware/base.rb]: I am nitpicking but I would avoid a temporary `response_headers` and just do `when .... .merge!(...)`.

**Comment by dblock** [comment]
Sorry more minor comments, please.

**Comment by namusyaka** [comment]
Thanks for polite reviewing! Pushed.

**Comment by dblock** [comment]
Merged, thanks!
